### DISCUSSION TITLE
few ideas

### DISCUSSION THREAD

**Discussion by josef2600** [pschatzmann/arduino-audio-tools#1044] [Q&A] [answered]
as i understand, all of your code runs on core 0. and Arduino runs everything on core 0 too. including all of WIFI codes and other libs. if possible, it is best to run some demanding codes on core 1. all esp32 (except c3-c6) are dual core, so why not use it? and esp32c3 is not going to be very good at it. but you can also add the changes for both of them. you are already have this section for setting different micro's , so why not adding core number to it too? for example, in file:
`arduino-audio-tools-main\src\AudioConfig.h`
in line 229 we can have this `#define URL_STREAM_CORE 1`
and why do you have `checkMemory(1)` by default? it just slow the program.
thank you.

**Comment by pschatzmann** [comment, marked as answer]
That's not correct: The Arduino loop runs on core 1 and all the networking is running on core 0.
checkMemory is only executed (added to the code) when USE_CHECK_MEMORY is set to true: By default this is off. 

URL_STREAM_CORE is only used by URLStreamBuffered which is usually not needed because the regular implementation is fast enough and I therefore do not even have any examples for it.

**Reply by josef2600** [reply]
sorry, i didn't notice `#define USE_CHECK_MEMORY 0` .
and yes, Arduino loop runs on core 1. thanks.

**Comment by josef2600** [comment]
in `arduino-audio-tools-main\src\AudioTimer\AudioTimerESP32.h`
in line 308 is best to add `taskYIELD();`
in line 322 you have `delayMicroseconds`. this is not good, it is a blocking code and i think it is better to replace it with something else like `asm("nop");` or something shorter. i dont know what it is for! but RTOS has no delay less than 1 ms. at least not for these micro's!
and again, in line 234is best to add `taskYIELD();`

**Reply by pschatzmann** [reply]
Are you planning to use this  ? 
From all the options this is the least preferred one and usually there is not really a good reason to do this. 
You have the choice between enum TimerFunction {DirectTimerCallback, TimerCallbackInThread, SimpleThreadLoop }; 
And by default the DirectTimerCallback is used. You would use TimerCallbackInThread when the callback is taking too long or is not supported by ESP32 and SimpleThreadLoop when you used up all your timers and you are with no other option left.

But thanks for you comments: I will update my code...

**Reply by pschatzmann** [reply]
ps. the Timer is currently nowhere used for Audio in the ESP32

**Reply by josef2600** [reply]
thank you. and yes i am using this way in all my codes. this causes minimum damage for running and most powerful way to do it. 
i sow you defined 3 timer in your code, but didn't look farther to see if they are used or not. why create them if you are not using them?

**Comment by josef2600** [comment]
in file `arduino-audio-tools-main\src\AudioTools\StreamCopy.h` in line 377 you have:
`size_t bytes_to_read = min(result, static_cast<size_t>(buffer_size) );`
not to be touchy (!) , but out of curiosity, isn't it better to remove `min` and make it like this:
```
if (result > buffer_size) result = buffer_size;
size_t bytes_to_read = result;
```
because esp32 or ARM's in general are really stupid (like their designers!) and don't have many simple calculations, like min max div or even minus (-)  !!  , this will be alot faster? and maybe ... maybe (!!) don't use int ??!! and only use uint32_t or uint16_t for things that dont have negative numbers? it would be faster! who knows? not me! i don't know! 
thanks.

**Reply by pschatzmann** [reply]
Did you try to messure the difference?
